src/shims/fs.rs

```diff
@@ -257,6 +264,186 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>, // Should be an `int`
+        pathname_op: OpTy<'tcx, Tag>, // Should be a `const char *`
+        flags_op: OpTy<'tcx, Tag>, // Should be an `int`
+        _mask_op: OpTy<'tcx, Tag>, // Should be an `unsigned int`
+        statxbuf_op: OpTy<'tcx, Tag> // Should be a `struct statx *`
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // Under normal circumstances, we would use `deref_operand(statxbuf_op)` to produce a
+        // proper `MemPlace` and then write the results of this function to it. However, the
+        // syscall API in Rust is untyped. This means that all the `statx` parameters are provided
+        // as `isize`s instead of having the proper types. Thus, we have to recover the layout of
+        // `statxbuf_op` by using the `libc::statx` struct type.
+        let statxbuf_place = {
+            // FIXME: This long path is required because `libc::statx` is an struct and also a
+            // function and `resolve_path` is returning the latter.
+            let statx_ty = this
+                .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
+                .ty(*this.tcx);
+            let statxbuf_ty = this.tcx.mk_mut_ptr(statx_ty);
+            let statxbuf_layout = this.layout_of(statxbuf_ty)?;
+            let statxbuf_imm = ImmTy::from_scalar(statxbuf_scalar, statxbuf_layout);
+            this.ref_to_mplace(statxbuf_imm)?
+        };
+
+        let pathname: PathBuf = this.read_os_str_from_c_str(pathname_scalar)?.into();
+        // This argument should be a `c_int` but the syscall API just returns a pointer sized
+        // integer.
+        let flags: i32 = this.try_from_ptr_sized_operand(flags_op)?;
```

Comment:
Where is that `statx` call site? I seems just wrong to me to pass an `isize` when the syscall says it should be a `c_int`.

Reply:
So the reason these are all `isize` is that they all [get cast right here](https://github.com/rust-lang/rust/blob/3982d3514efbb65b3efac6bb006b3fa496d16663/src/libstd/sys/unix/weak.rs#L92). That is very odd.

@gnzlbg any idea why libstd turns all syscall parameters to `c_long`, even when the official syscall docs use other types such as "int" or "unsinged"?

Reply:
Ehm, no idea, the first argument to `glibc`'s syscall is a `c_long` specifying the number of arguments, but the rest of the arguments are a C variadic list of arguments, and they should have the correct types of the syscall. If the syscall expects an `c_int`, and you pass it an `isize`, then that would have the wrong argument size on, e.g., 64-bit linux, so that seems wrong. Maybe it doesn't "matter" in practice, e.g., due to how the calling convention handles these arguments, but still, looks like a bug. cc @amanieu

Reply:
The first argument seems to be the syscall number, not the number of arguments?

> If the syscall expects an c_int, and you pass it an isize, then that would have the wrong argument size on, e.g., 64-bit linux, so that seems wrong.

I think that is exactly what is happening. All parameters of all syscalls going through that macro are passed as `c_long`, no matter their actual type.